csrc/scheduler/normalization.cpp

```diff
@@ -294,6 +294,64 @@ std::shared_ptr<ReductionParams> innerOuterPersistentHeuristic(
 }
 // Copied from reduction scheduler, should generalize. Simply needed to take out
 // grid reductions.
+std::shared_ptr<ReductionParams> innerPersistentHeuristicSharedMemory(
```

Comment:
Please add assertions to validate the assumptions. For example, looks like `total_reduction_numel` must be larger than `vectorize_factor * bdimx`. 

Doesn't using the existing inner heuristic but just with shared memory work well?

Reply:
It works well. I added another function because: (1) the original function is already very long.  (2) a seperated function provides a short pass which reduces overhead. (3) It would be easier for future maintaince in case we want to do more optimizations for smem persistent.

Reply:
I agree with (1). For (2), I don't think the difference is noticeable. For (3), I guess that would depend how similar the two schedulers are. If they are mostly the same except for the memory types, then we should just have one scheduler for both cases. If the shared-mem case is much more different, I'd make more sense to have a separate implementation

Reply:
That said, this is not a blocker.